hie-build-plugin/Haskell/Ide/BuildPlugin.hs

```diff
@@ -100,21 +111,91 @@ buildPluginDescriptor = PluginDescriptor
   , pdExposedServices = []
   , pdUsedServices    = []
   }
+-}
+
+buildPluginDescriptor :: PluginDescriptor
+buildPluginDescriptor = PluginDescriptor
+  {
+    pluginName = "Build plugin"
+  , pluginDesc = "A HIE plugin for building cabal/stack packages"
+  , pluginCommands =
+      [ PluginCommand "prepare"
+                      "Prepares helper executable. The project must be configured first"
+                      prepareHelper
+      -- , PluginCommand "isPrepared"
+      --                    ("Checks whether cabal-helper is prepared to work with this project. "
+      --                  <> "The project must be configured first")
+      --                  isHelperPrepared
+      , PluginCommand "isConfigured"
+                       "Checks if project is configured"
+                       isConfigured
+      , PluginCommand "configure"
+                         ("Configures the project. "
+                       <> "For stack project with multiple local packages - build it")
+                       configure
+      , PluginCommand "listTargets"
+                      "Given a directory with stack/cabal project lists all its targets"
+                      listTargets
+      , PluginCommand "listFlags"
+                      "Lists all flags that can be set when configuring a package"
+                      listFlags
+      , PluginCommand "buildDirectory"
+                      "Builds all targets that correspond to the specified directory"
+                      buildDirectory
+      , PluginCommand "buildTarget"
+                      "Builds specified cabal or stack component"
+                      buildTarget
+      ]
+  }
 
 data OperationMode = StackMode | CabalMode
 
 readMode "stack" = Just StackMode
 readMode "cabal" = Just CabalMode
 readMode _ = Nothing
 
+-- | Used internally by commands, all fields always populated, possibly with
+-- default values
 data CommonArgs = CommonArgs {
          caMode :: OperationMode
         ,caDistDir :: String
         ,caCabal :: String
         ,caStack :: String
     }
 
-withCommonArgs ctx req a = do
+-- | Used to interface with the transport, where the mode is required but rest
+-- are optional
+data CommonParams = CommonParams {
+         cpMode    :: T.Text
+        ,cpDistDir :: Maybe String
+        ,cpCabal   :: Maybe String
+        ,cpStack   :: Maybe String
+        ,cpFile    :: Uri
+    } deriving Generic
+
+instance FromJSON CommonParams where
+  parseJSON = J.genericParseJSON $ customOptions 2
+instance ToJSON CommonParams where
+  toJSON = J.genericToJSON $ customOptions 2
+
+incorrectParameter = undefined
```

Comment:
yes. I forgot about that, to be honest.

This is very much a "smallest effort to get it to compile" version